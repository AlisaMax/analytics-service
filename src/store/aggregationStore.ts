import { create } from 'zustand';
import type { AggregationResult } from '../types/aggregationResult';

interface AggregationState {
  result: AggregationResult | null;
  setResult: (newResult: AggregationResult) => void;
}

export const useAggregationStore = create<AggregationState>((set) => ({
  result: null,
  setResult: (newResult) => set({ result: newResult }),
}));
