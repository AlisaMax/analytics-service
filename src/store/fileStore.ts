import { create } from 'zustand';

export enum FileStatus {
  INITIAL = 'initial',
  UPLOADING = 'uploading',
  SUCCESS = 'success',
  ERROR = 'error',
  LOADING = 'loading',
  ADDED = 'added',
}

interface FileStore {
  file: File | null;
  isCorrect: boolean;
  status: FileStatus;
  setFile: (file: File | null) => void;
  setIsCorrect: (value: boolean) => void;
  setStatus: (value: FileStatus) => void;
}

export const useFileStore = create<FileStore>((set) => ({
  file: null,
  isCorrect: true,
  status: FileStatus.INITIAL,

  setFile: (file) => set({ file }),
  setIsCorrect: (value) => set({ isCorrect: value }),
  setStatus: (value) => set({ status: value }),
}));
