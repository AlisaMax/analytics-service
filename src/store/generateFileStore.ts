import { create } from 'zustand';

export enum GenerateFileStatus {
  INITIAL = 'initial',
  LOADING = 'loading',
  SUCCESS = 'success',
  ERROR = 'error',
}

interface GenerateFileStore {
  status: GenerateFileStatus;
  setStatus: (value: GenerateFileStatus) => void;
}

export const useGenerateFileStore = create<GenerateFileStore>((set) => ({
  status: GenerateFileStatus.INITIAL,

  setStatus: (value) => set({ status: value }),
}));
