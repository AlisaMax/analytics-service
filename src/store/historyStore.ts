import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import type { AggregationResult } from '../types/aggregationResult';

interface HistoryItem {
  id: string;
  name: string;
  date: string;
  success: boolean;
  result?: AggregationResult;
}
interface HistoryStore {
  history: HistoryItem[];
  addHistoryItem: (item: HistoryItem) => void;
  removeHistoryItemById: (id: string) => void;
  clearHistory: () => void;
}

export const useHistoryStore = create<HistoryStore>()(
  persist(
    (set, get) => ({
      history: [],
      addHistoryItem: (item) => set({ history: [...get().history, item] }),

      removeHistoryItemById: (id) =>
        set({ history: get().history.filter((item) => item.id !== id) }),

      clearHistory: () => set({ history: [] }),
    }),
    {
      name: 'aggregationHistory',
    }
  )
);
