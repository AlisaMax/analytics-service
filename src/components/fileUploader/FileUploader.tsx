import React, { useRef } from 'react';
import styles from './FileUploader.module.css';
import cn from 'classnames';
import Button from '../UI/button/Button';
import {
  handleDrop,
  onDragEnter,
  onDragLeave,
  onDragOver,
} from '../../features/fileUploader/DragAndDrop';
import { useFileStore, FileStatus } from '../../store/fileStore';
import { HandlerUpload } from '../../features/fileUploader/HandleUpload';
import Loader from '../UI/loader/Loader';
import { handleFileChange } from '../../features/fileUploader/HandleFileChange';
import ActionResultButton from '../UI/actionResultButton/ActionResultButton';

const FileUploader = () => {
  const status = useFileStore((state) => state.status);
  const file = useFileStore((state) => state.file);
  const inputRef = useRef<HTMLInputElement | null>(null);

  const renderHelpText = (): string => {
    switch (status) {
      case FileStatus.INITIAL:
      case FileStatus.UPLOADING:
        return 'или перетащите сюда';
      case FileStatus.ERROR:
        return 'упс, не то...';
      case FileStatus.ADDED:
        return 'файл загружен!';
      case FileStatus.LOADING:
        return 'идёт парсинг файла';
      case FileStatus.SUCCESS:
        return 'готово!';
      default:
        return '';
    }
  };

  const renderUploader = (): React.ReactElement | string => {
    switch (status) {
      case FileStatus.INITIAL:
      case FileStatus.UPLOADING:
        return (
          <label className={styles.inputFileButton} htmlFor="inputFile">
            <input
              id="inputFile"
              ref={inputRef}
              className={styles.InputFile}
              multiple={false}
              type="file"
              onClick={() => useFileStore.getState().setStatus(FileStatus.UPLOADING)}
              onChange={(e) => handleFileChange(e)}
            />
            {(() => {
              if (status === FileStatus.INITIAL || status === FileStatus.UPLOADING)
                return 'Загрузить файл';
              return file?.name || '';
            })()}
          </label>
        );
      case FileStatus.ERROR:
      case FileStatus.SUCCESS:
      case FileStatus.ADDED:
        return <ActionResultButton title={file?.name} status={status} inputRef={inputRef} />;
      case FileStatus.LOADING:
        return <Loader />;
      default:
        return '';
    }
  };

  return (
    <div className={cn(styles.UplodeWrap, styles[status])}>
      <div
        className={cn(styles.UplodeBlock)}
        onDrop={handleDrop}
        onDragOver={onDragOver}
        onDragEnter={onDragEnter}
        onDragLeave={onDragLeave}
      >
        <div className={styles.fileAddWrap}>{renderUploader()}</div>
        <p className={styles.HelpText}>{renderHelpText()}</p>
      </div>
      <Button
        label="Отправить"
        isDisabled={status !== FileStatus.ADDED}
        className={styles.sendButton}
        onClick={HandlerUpload}
      />
    </div>
  );
};

export default FileUploader;
