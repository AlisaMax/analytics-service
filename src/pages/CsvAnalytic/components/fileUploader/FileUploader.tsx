import cn from 'classnames';
import styles from './FileUploader.module.css';
import Button from '../../../../components/button/Button';
import { handleDrop, onDragEnter, onDragLeave, onDragOver } from './utils/DragAndDrop';
import { useFileStore, FileStatus } from '../../../../store/fileStore';
import { aggregateFile } from '../../../../features/fileUploader/aggregateFile';
import ActionResultButton from '../../../../components/actionResultButton/ActionResultButton';
import Loader from '../../../../components/loader/Loader';
import { handleFileChange } from '../../../../features/fileUploader/handleFileChange';
import { useRef } from 'react';

const FileUploader = () => {
  const status = useFileStore((state) => state.status);
  const file = useFileStore((state) => state.file);
  const setStatus = useFileStore((state) => state.setStatus);
  const inputRef = useRef<HTMLInputElement | null>(null);

  const renderByStatus = (): { uploader: React.ReactNode; helpText: string } => {
    const baseHelpText = 'или перетащите сюда';

    switch (status) {
      case FileStatus.INITIAL:
      case FileStatus.UPLOADING:
        return {
          helpText: baseHelpText,
          uploader: (
            <label className={styles.inputFileButton} htmlFor="inputFile" data-testid="input-file">
              <input
                id="inputFile"
                ref={inputRef}
                className={styles.InputFile}
                multiple={false}
                type="file"
                onClick={() => setStatus(FileStatus.UPLOADING)}
                onChange={handleFileChange}
              />
              {status === FileStatus.INITIAL || status === FileStatus.UPLOADING
                ? 'Загрузить файл'
                : file?.name || ''}
            </label>
          ),
        };

      case FileStatus.ADDED:
        return {
          helpText: 'файл загружен!',
          uploader: <ActionResultButton title={file?.name} status={status} inputRef={inputRef} />,
        };

      case FileStatus.LOADING:
        return {
          helpText: 'идёт парсинг файла',
          uploader: <Loader />,
        };

      case FileStatus.SUCCESS:
        return {
          helpText: 'готово!',
          uploader: <ActionResultButton title={file?.name} status={status} inputRef={inputRef} />,
        };

      case FileStatus.ERROR:
        return {
          helpText: 'упс, не то...',
          uploader: <ActionResultButton title={file?.name} status={status} inputRef={inputRef} />,
        };

      default:
        return {
          helpText: '',
          uploader: '',
        };
    }
  };

  const { uploader, helpText } = renderByStatus();

  return (
    <div className={cn(styles.UplodeWrap, styles[status])}>
      <div
        className={cn(styles.UplodeBlock)}
        onDrop={handleDrop}
        onDragOver={onDragOver}
        onDragEnter={onDragEnter}
        onDragLeave={onDragLeave}
      >
        <div className={styles.fileAddWrap}>{uploader}</div>
        <p data-testid="help-text" className={styles.HelpText}>
          {helpText}
        </p>
      </div>
      <Button
        label="Отправить"
        isDisabled={status !== FileStatus.ADDED}
        className={styles.sendButton}
        onClick={aggregateFile}
        data-testid="send-button"
      />
    </div>
  );
};

export default FileUploader;
