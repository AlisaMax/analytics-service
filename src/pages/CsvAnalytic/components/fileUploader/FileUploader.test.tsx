import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import { FileStatus, useFileStore } from '../../../../store/fileStore';
import FileUploader from './FileUploader';
import { beforeEach } from 'node:test';

vi.mock('../../../../store/fileStore');

describe('Загрузка файла', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it.each(Object.values(FileStatus))('рендерит компонент в зависимости от статуса %s', (status) => {
    const mockSetStatus = vi.fn();
    const mockSetFile = vi.fn();

    const mockStore = {
      file: null,
      status: status,
      setFile: mockSetFile,
      setStatus: mockSetStatus,
    };

    const mockedUseFileStore = vi.mocked(useFileStore);
    mockedUseFileStore.mockImplementation((selector) => {
      return selector(mockStore);
    });

    const { asFragment } = render(<FileUploader />);

    expect(asFragment()).toMatchSnapshot();
  });
});
