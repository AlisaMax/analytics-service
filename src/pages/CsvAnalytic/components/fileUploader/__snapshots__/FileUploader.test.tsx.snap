// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Загрузка файла > рендерит компонент в зависимости от статуса added 1`] = `
<DocumentFragment>
  <div
    class="_UplodeWrap_a550f8 _added_a550f8"
  >
    <div
      class="_UplodeBlock_a550f8"
    >
      <div
        class="_fileAddWrap_a550f8"
      >
        <div
          class="_actionResultWrapper_695e79 _added_695e79"
        >
          <div
            class="_BlockTitle_695e79"
          />
          <span
            class="_clearFileButton_695e79"
          >
            <img
              alt="очистить"
              src="data:image/svg+xml,%3csvg%20width='33'%20height='32'%20viewBox='0%200%2033%2032'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M7.16675%2025.3333L16.5001%2016M16.5001%2016L25.8334%206.66663M16.5001%2016L7.16675%206.66663M16.5001%2016L25.8334%2025.3333'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/svg%3e"
            />
          </span>
        </div>
      </div>
      <p
        class="_HelpText_a550f8"
      >
        файл загружен!
      </p>
    </div>
    <button
      class="_sendButton_a550f8"
    >
      Отправить
    </button>
  </div>
</DocumentFragment>
`;

exports[`Загрузка файла > рендерит компонент в зависимости от статуса error 1`] = `
<DocumentFragment>
  <div
    class="_UplodeWrap_a550f8 _error_a550f8"
  >
    <div
      class="_UplodeBlock_a550f8"
    >
      <div
        class="_fileAddWrap_a550f8"
      >
        <div
          class="_actionResultWrapper_695e79 _error_695e79"
        >
          <div
            class="_BlockTitle_695e79"
          />
          <span
            class="_clearFileButton_695e79"
          >
            <img
              alt="очистить"
              src="data:image/svg+xml,%3csvg%20width='33'%20height='32'%20viewBox='0%200%2033%2032'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M7.16675%2025.3333L16.5001%2016M16.5001%2016L25.8334%206.66663M16.5001%2016L7.16675%206.66663M16.5001%2016L25.8334%2025.3333'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/svg%3e"
            />
          </span>
        </div>
      </div>
      <p
        class="_HelpText_a550f8"
      >
        упс, не то...
      </p>
    </div>
    <button
      class="_sendButton_a550f8"
      disabled=""
    >
      Отправить
    </button>
  </div>
</DocumentFragment>
`;

exports[`Загрузка файла > рендерит компонент в зависимости от статуса initial 1`] = `
<DocumentFragment>
  <div
    class="_UplodeWrap_a550f8 _initial_a550f8"
  >
    <div
      class="_UplodeBlock_a550f8"
    >
      <div
        class="_fileAddWrap_a550f8"
      >
        <label
          class="_inputFileButton_a550f8"
          for="inputFile"
        >
          <input
            class="_InputFile_a550f8"
            id="inputFile"
            type="file"
          />
          Загрузить файл
        </label>
      </div>
      <p
        class="_HelpText_a550f8"
      >
        или перетащите сюда
      </p>
    </div>
    <button
      class="_sendButton_a550f8"
      disabled=""
    >
      Отправить
    </button>
  </div>
</DocumentFragment>
`;

exports[`Загрузка файла > рендерит компонент в зависимости от статуса loading 1`] = `
<DocumentFragment>
  <div
    class="_UplodeWrap_a550f8 _loading_a550f8"
  >
    <div
      class="_UplodeBlock_a550f8"
    >
      <div
        class="_fileAddWrap_a550f8"
      >
        <div
          class="_LoadingBlock_f82386"
        >
          <div
            class="_Loading_f82386"
          />
        </div>
      </div>
      <p
        class="_HelpText_a550f8"
      >
        идёт парсинг файла
      </p>
    </div>
    <button
      class="_sendButton_a550f8"
      disabled=""
    >
      Отправить
    </button>
  </div>
</DocumentFragment>
`;

exports[`Загрузка файла > рендерит компонент в зависимости от статуса success 1`] = `
<DocumentFragment>
  <div
    class="_UplodeWrap_a550f8 _success_a550f8"
  >
    <div
      class="_UplodeBlock_a550f8"
    >
      <div
        class="_fileAddWrap_a550f8"
      >
        <div
          class="_actionResultWrapper_695e79 _success_695e79"
        >
          <div
            class="_BlockTitle_695e79"
          />
          <span
            class="_clearFileButton_695e79"
          >
            <img
              alt="очистить"
              src="data:image/svg+xml,%3csvg%20width='33'%20height='32'%20viewBox='0%200%2033%2032'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M7.16675%2025.3333L16.5001%2016M16.5001%2016L25.8334%206.66663M16.5001%2016L7.16675%206.66663M16.5001%2016L25.8334%2025.3333'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/svg%3e"
            />
          </span>
        </div>
      </div>
      <p
        class="_HelpText_a550f8"
      >
        готово!
      </p>
    </div>
    <button
      class="_sendButton_a550f8"
      disabled=""
    >
      Отправить
    </button>
  </div>
</DocumentFragment>
`;

exports[`Загрузка файла > рендерит компонент в зависимости от статуса uploading 1`] = `
<DocumentFragment>
  <div
    class="_UplodeWrap_a550f8 _uploading_a550f8"
  >
    <div
      class="_UplodeBlock_a550f8"
    >
      <div
        class="_fileAddWrap_a550f8"
      >
        <label
          class="_inputFileButton_a550f8"
          for="inputFile"
        >
          <input
            class="_InputFile_a550f8"
            id="inputFile"
            type="file"
          />
          Загрузить файл
        </label>
      </div>
      <p
        class="_HelpText_a550f8"
      >
        или перетащите сюда
      </p>
    </div>
    <button
      class="_sendButton_a550f8"
      disabled=""
    >
      Отправить
    </button>
  </div>
</DocumentFragment>
`;
